src/exact/ExactMapper.cpp

```diff
@@ -635,13 +643,13 @@ void ExactMapper::coreMappingRoutine(const std::set<unsigned short>& qubitChoice
                 if (gate.singleQubit())
                     continue;
 
-                expr reverse = c.bool_val(true);
+                LogicTerm reverse = LogicTerm(true);
                 for (const auto& edge: rcm) {
                     auto indexFT = x[k][physicalQubitIndex[edge.first]][gate.target];
                     auto indexSC = x[k][physicalQubitIndex[edge.second]][gate.control];
                     reverse      = reverse && (!indexFT || !indexSC);
                 }
-                opt.add(reverse.simplify(), GATES_OF_DIRECTION_REVERSE);
+                dynamic_cast<LogicBlockOptimizer*>(lb)->weightedTerm(reverse, GATES_OF_DIRECTION_REVERSE);
```

Comment:
Is is guaranteed that this dynamic cast is successful? Shouldn't there be a check? otherwise `lb` could just straight-up be a `LogicBlockOptimizer`. Then there is no need for a cast at all.